src/Microsoft.VisualStudio.Validation/Requires.cs

```diff
@@ -418,6 +420,72 @@ public static void Argument([DoesNotReturnIf(false)] bool condition, string? par
+        /// <param name="resourceName">The name of the string resource to obtain for the exception message. For example: <c>nameof(Strings.SomeError)</c>.</param>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="resourceManager"/> is <see langword="null"/>.</exception>
+        /// <exception cref="ArgumentException">Thrown if <paramref name="condition"/> is <see langword="false"/>.</exception>
+        /// <remarks>
+        /// This overload allows only loading a localized string in the error condition as an optimization in perf critical sections over the simpler
+        /// to use <see cref="Argument(bool, string?, string?)"/> overload.
+        /// </remarks>
+        [DebuggerStepThrough]
+        public static void Argument([DoesNotReturnIf(false)] bool condition, string? parameterName, ResourceManager resourceManager, string resourceName)
+        {
+            NotNull(resourceManager, nameof(resourceManager));
+            if (!condition)
+            {
+                throw new ArgumentException(resourceManager.GetString(resourceName, CultureInfo.CurrentCulture), parameterName);
+            }
+        }
+
+        /// <inheritdoc cref="Argument(bool, string?, ResourceManager, string, object?, object?)"/>
+        public static void Argument([DoesNotReturnIf(false)] bool condition, string? parameterName, ResourceManager resourceManager, string unformattedMessageResourceName, object? arg1)
+        {
+            NotNull(resourceManager, nameof(resourceManager));
+            if (!condition)
+            {
+                throw new ArgumentException(Format(resourceManager.GetString(unformattedMessageResourceName, CultureInfo.CurrentCulture), arg1), parameterName);
+            }
+        }
+
+        /// <inheritdoc cref="Argument(bool, string?, ResourceManager, string)"/>
+        /// <param name="unformattedMessageResourceName">The name of the string resource to obtain for the exception message. For example: <c>nameof(Strings.SomeError)</c>.</param>
+        /// <param name="condition"><inheritdoc cref="Argument(bool, string?, ResourceManager, string)" path="/param[@name='condition']"/></param>
+        /// <param name="parameterName"><inheritdoc cref="Argument(bool, string?, ResourceManager, string)" path="/param[@name='parameterName']"/></param>
+        /// <param name="resourceManager"><inheritdoc cref="Argument(bool, string?, ResourceManager, string)" path="/param[@name='resourceManager']"/></param>
+        /// <param name="arg1">The first formatting argument.</param>
+        /// <param name="arg2">The second formatting argument.</param>
+        [DebuggerStepThrough]
+        public static void Argument([DoesNotReturnIf(false)] bool condition, string? parameterName, ResourceManager resourceManager, string unformattedMessageResourceName, object? arg1, object? arg2)
```

Comment:
Can we not utilize the "params object[] args" syntax here to support arbitrary number of args?

Reply:
We do have the overload you propose after this one. Yes, the `arg` and `arg1, arg2` overloads are declared to avoid the array allocation for common cases with few args.